### PR TITLE
Modernize aws_sns_topic_subscription

### LINKED ISSUE #12286: Support email sns_topic_subscription
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Please support email protocol for sns_topic_subscription. I understand that creation is not possible (as per note in docs, see below in references). But existing topics can be imported (that works already). The use case is - for TF destroy and for monitoring changes (with terraform plan - vs any modifications done outside of TF) and applying changes that don't need resource re-creation. If the subscription is in non-confirmed state, then import should fail, preventing inability of TF to destroy or modify unconfirmed resource.

I understand that this is limited functionality, but still very valuable.

### New or Affected Resource(s)

sns_topic_subscription

### Potential Terraform Configuration

```hcl
resource "aws_sns_topic_subscription" "abc" {
  endpoint             = "abc@def.com"
  protocol             = "email"
  raw_message_delivery = "false"
  topic_arn            = aws_sns_topic.abc.id
}.
```

### References

https://github.com/terraform-providers/terraform-provider-aws/issues/7782

Docs: https://www.terraform.io/docs/providers/aws/r/sns_topic_subscription.html
> Unsupported protocols include the following:
>
> email -- delivery of message via SMTP
> email-json -- delivery of JSON-encoded message via SMTP
> These are unsupported because the endpoint needs to be authorized and does not generate an ARN until the target email address has been validated. This breaks the Terraform model and as a result are not currently supported.

### LINKED ISSUE #17303: sns_topic_subscription support firehose protocol
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS recently released the ability to subscribe kinesis firehose to an SNS topic. Please add the ability to subscribe `firehose` protocol to SNS topic.

### New or Affected Resource(s)

* sns_topic_subscription

### Potential Terraform Configuration

```hcl
resource "aws_sns_topic_subscription" "example" {
  topic_arn = aws_sns_topic.example.arn
  protocol  = "firehose"
  endpoint  = "arn:aws:firehose:us-west-2:123456789012:deliverystream/MyStream"
  subscription_role_arn = "arn:aws:iam::123456789012:role/FirehoseRoleArn"
}
```

### References
https://docs.aws.amazon.com/sns/latest/api/API_Subscribe.html#API_Subscribe_RequestParameters

* #0000

### PR DESCRIPTION
**Pull request by nikhil-goenka**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12286, closes #17303, closes #17307 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note

```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by nikhil-goenka** [comment]
@bflad : Just an approach for email support on SNS, what are your thoughts ?

File: website/docs/r/sns_topic_subscription.html.markdown [hunk 1]
```diff
@@ -248,17 +248,14 @@ Supported SNS protocols include:
...
 * `https` -- delivery of JSON-encoded messages via HTTPS. Supported only for the end points that auto confirms the subscription.
 
-Unsupported protocols include the following:
-
-* `email` -- delivery of message via SMTP
-* `email-json` -- delivery of JSON-encoded message via SMTP
-
 These are unsupported because the endpoint needs to be authorized and does not
 generate an ARN until the target email address has been validated. This breaks
 the Terraform model and as a result are not currently supported.
```
Comment (dmccaffery) [inline website/docs/r/sns_topic_subscription.html.markdown:261]: Should this paragraph also be removed as part of this PR?

**Comment by okgolove** [comment]
Any chances this one will be merged?

**Comment by YakDriver** [comment]
@nikhil-goenka Thank you for your work on this PR! I will be working on this before long. In order to expedite the process, I will likely make some changes. Make sure that you have checked the box "Allow edits from maintainers." (Don't worry though, you will receive all credit for your contribution and code!) Also, please coordinate with us before making any commits to this branch. Again, thank you for your help and we look forward to this popular addition to the AWS provider!

**Comment by nikhil-goenka** [comment]
> @nikhil-goenka Thank you for your work on this PR! I will be working on this before long. In order to expedite the process, I will likely make some changes. Make sure that you have checked the box "Allow edits from maintainers." (Don't worry though, you will receive all credit for your contribution and code!) Also, please coordinate with us before making any commits to this branch. Again, thank you for your help and we look forward to this popular addition to the AWS provider!

Thank you and will be great to see this branch getting merged, the box has been ticked.

**Comment by YakDriver** [comment]
The biggest problem with email support is that unconfirmed subscriptions cannot be deleted / unsubscribed. My workaround is to remove the `aws_sns_topic_subscription` from the state, log a warning, and provide plenty of documentation for this issue. I'd love to get feedback on this approach.

**Review by YakDriver — APPROVED** [review]
Looks great! Thanks for your work @nikhil-goenka and @chammock! 🎉 

Acceptance tests in GovCloud:
```
--- SKIP: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (1.28s)
--- SKIP: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (1.28s)
--- PASS: TestAccAWSSNSTopicSubscription_email (18.14s)
--- PASS: TestAccAWSSNSTopicSubscription_basic (21.83s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (47.23s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (48.52s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (48.52s)
--- PASS: TestAccAWSSNSTopicSubscription_redrivePolicy (52.97s)
--- SKIP: TestAccAWSSNSTopicSubscription_firehose (53.93s)
``` 

On `us-west-2`:
```
--- PASS: TestAccAWSSNSTopicSubscription_email (14.93s)
--- PASS: TestAccAWSSNSTopicSubscription_basic (18.18s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (37.10s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (37.97s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (38.05s)
--- PASS: TestAccAWSSNSTopicSubscription_redrivePolicy (40.86s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (41.26s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (81.65s)
--- PASS: TestAccAWSSNSTopicSubscription_firehose (100.05s)
```

**Pull request merged**